tests/unittests/HyphenTest.cpp

```diff
@@ -0,0 +1,322 @@
+const vector<const char *> TrainingData{
+    "ad-mi-ni-stra-tion",
+    "ad-mit",
+    "al-low",
+    "al-though",
+    "an-i-mal",
+    "any-one",
+    "ar-rive",
+    "art",
+    "at-tor-ney",
+    "be-cause",
+    "be-fore",
+    "be-ha-vior",
+    "can-cer",
+    "cer-tain-ly",
+    "con-gress",
+    "coun-try",
+    "cul-tural",
+    "cul-ture",
+    "de-cide",
+    "de-fense",
+    "de-gree",
+    "de-sign",
+    "de-spite",
+    "de-velop",
+    "di-rec-tion",
+    "di-rec-tor",
+    "dis-cus-sion",
+    "eco-nomy",
+    "elec-tion",
+    "en-vi-ron-men-tal",
+    "es-tab-lish",
+    "ev-ery-one",
+    "ex-actly",
+    "ex-ec-u-tive",
+    "ex-ist",
+    "ex-pe-ri-ence",
+    "ex-plain",
+    "fi-nally",
+    "for-get",
+    "hun-dred",
+    "in-crease",
+    "in-di-vi-du-al",
+    "it-self",
+    "lan-guage",
+    "le-gal",
+    "lit-tle",
+    "lo-cal",
+    "ma-jo-ri-ty",
+    "ma-te-rial",
+    "may-be",
+    "me-di-cal",
+    "meet-ing",
+    "men-tion",
+    "mid-dle",
+    "na-tion",
+    "na-tional",
+    "oc-cur",
+    "of-fi-cer",
+    "par-tic-u-lar-ly",
+    "pat-tern",
+    "pe-riod",
+    "phy-si-cal",
+    "po-si-tion",
+    "pol-icy",
+    "pos-si-ble",
+    "pre-vent",
+    "pres-sure",
+    "pro-per-ty",
+    "pur-pose",
+    "re-cog-nize",
+    "re-gion",
+    "re-la-tion-ship",
+    "re-main",
+    "re-sponse",
+    "re-sult",
+    "rea-son",
+    "sea-son",
+    "sex-ual",
+    "si-mi-lar",
+    "sig-ni-fi-cant",
+    "sim-ple",
+    "sud-den-ly",
+    "sum-mer",
+    "thou-sand",
+    "to-day",
+    "train-ing",
+    "treat-ment",
+    "va-ri-ous",
+    "value",
+    "vi-o-lence",
+};
+
+TEST(HyphenTest, network) {
+  // Convert the training data to word windows and labels.
+  vector<string> words;
+  vector<bool> hyphens;
+  for (auto *hword : TrainingData) {
+    dehyphenate(hword, words, hyphens);
+  }
+
+  // This depends on the training data, of course.
+  const size_t numSamples = 566;
+  ASSERT_EQ(hyphens.size(), numSamples);
+  ASSERT_EQ(words.size(), numSamples);
+
+  // Randomly shuffle the training data.
+  // This is required for stochastic gradient descent training.
+  for (size_t i = 0; i != numSamples; i++) {
+    size_t j = nextRandInt(0, numSamples - 1);
+    std::swap(words[i], words[j]);
+    std::swap(hyphens[i], hyphens[j]);
+  }
+
+  // Convert words and hyphens to a tensor representation.
+  Tensor inputs(ElemKind::FloatTy, {numSamples, 6, 27});
+  Tensor outputs(ElemKind::FloatTy, {numSamples, 1});
+  auto inputHandle = inputs.getHandle<float>();
+  auto outputHandle = outputs.getHandle<float>();
+  for (size_t i = 0; i != numSamples; i++) {
+    mapLetterWindow(words[i], i, inputHandle);
+    outputHandle.at({i, 0}) = hyphens[i];
+  }
+
+  // Now build the network:
+  //
+  // Input: 6 x 27.
+  // Hidden: 16 fully connected, RELU output.
+  // Output: 1 fully connected, sigmoid output for logistic regression.
+  const size_t batchSize = numSamples;
```

Comment:
If `batchSize` is smaller, we can expect the network to converge to better accuracy and run faster. Let's try 5 or 15.

Reply:
Yeah, that would be nice. I couldn't figure out how to run SGD with mini-batches using the `EE.runBatch()` function. Does it support mini-batches automatically, or should I call it with `iterations=1` for each mini-batch?

Reply:
SGD should work fine. Please take a look at `nonLinearClassifier` unit test. Maybe you didn't specify `EE_.getConfig().batchSize`, and it didn't converge because of doing too big steps?